### PR TITLE
ROS2 build on QNX instructions review

### PR DESCRIPTION
**Pull request by asobhy-qnx**
ROS2 build on QNX instructions review

### CODE REVIEW

**Comment by asobhy-qnx** [comment]
@carlossvg, @LanderU here is the PR please feel free to add your feedback here.

**Comment by asobhy-qnx** [comment]
I will paste all the current feedback from the email here so that I do not miss anything

**Comment by asobhy-qnx** [comment]
Lander's feedback:

1-I'm using QNX 7.1, for some reason I've to copy "asio.hpp" into the rootfs otherwise Fast-DDS fails.

2-rmw_cyclonedds_cpp failed to build

3-Connected to point 3, I see that you're using rolling as a target, and setting to a certain git commit https://gitlab.com/qnx/ros2/cyclonedds_vendor/-/blob/master/CMakeLists.txt#L28. This is a bad idea for rolling release because the rest of the packages are pointing to master and this can be easily broken.

4-Connected to 2-3, in order to ensure that rolling builds for QNX I recommend setting up a CI that may be on https://gitlab.com/qnx. Which builds every night the latest changes to ensure the QNX build still works and if not, we are able to detect it and fix it. (Happy to help on this if you need any assistance). I recommend having a branch per each ROS 2 distro. And ideally a CI per each ROS 2 distro. This will be at least necessary for rolling.

5-Regarding CMake 3.18.0, is it possible to add a flag in the toolchain to avoid the update? Ideally, we should keep the CMake version used to build the rolling distro in the OR build farm.

This is the rmw_cyclone_dds error. This is because in Cyclone DDS this is not generated for version 509e499ee942025ed8d71e680d2ea9309528066c

```
/home/lander.usategui/ros2_rolling/src/ros2/rmw_cyclonedds/rmw_cyclonedds_cpp/src/rmw_node.cpp:80:10: fatal error: dds/ddsc/dds_data_allocator.h: No such file or directory
 #include "dds/ddsc/dds_data_allocator.h"

lander.usategui@qnx-xcompilation:~/ros2_rolling/build/x86_64/cyclonedds_vendor/cyclonedds_install/include/ddsc$ ls
dds.h
```

**Comment by carlossvg** [comment]
@asobhy-qnx We should add the document reference to the install instructions index

File: source/Installation/Rolling/QNX-Development-Setup.rst [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+   :depth: 2
+   :local:
+
+Overview of the build process
+-----------------------------
+
+Starting with a QNX SDP7.1 installation along with the required cross compiled dependencies, the build process will cross compile ROS 2's source code against SDP7.1 and the cross compiled dependencies.
+Binaries will be generated for the two architectures below:
+
+aarch64le
```
Comment (carlossvg) [inline source/Installation/Rolling/QNX-Development-Setup.rst:16]: @asobhy-qnx This is not correctly formatted as a list.

File: source/Installation/Rolling/QNX-Development-Setup.rst [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+Get ROS 2 code
+--------------
+
+Create a workspace and clone all repos:
+
+.. code-block:: bash
+
+   mkdir -p ~/ros2_rolling/src
+   cd ~/ros2_rolling
+   wget https://raw.githubusercontent.com/ros2/ros2/master/ros2.repos
```
Comment (carlossvg) [inline source/Installation/Rolling/QNX-Development-Setup.rst:121]: This should be: `wget https://raw.githubusercontent.com/ros2/ros2/rolling/ros2.repos`

File: source/Installation/Rolling/QNX-Development-Setup.rst [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+    cp ~/ros2_rolling/platform/build-ros2.sh ~/my_new_ws/
+
+Source your development environment which includes QNX environment and ROS2:
+
+.. code-block:: bash
+
+    . ~/qnx710/qnxsdp-env-ros2.sh
+    . ~/ros2_rolling/install/<your_target_arch>/local_setup.bash
+
+Create a package according to ttps://docs.ros.org/en/foxy/Tutorials/Creating-Your-First-ROS2-Package.html:
```
Comment (carlossvg) [inline source/Installation/Rolling/QNX-Development-Setup.rst:303]: The link is broken. Also I would use the same ROS release as the branch the instructions are. rolling in this case:

- https://docs.ros.org/en/rolling/Tutorials/Creating-Your-First-ROS2-Package.html

File: source/Installation/Rolling/QNX-Development-Setup.rst [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+    mkdir ~/my_new_ws/src
+    cd ~/my_new_ws/src
+    ros2 pkg create --build-type ament_cmake <package_name>
+
+
+After you write your code and are ready to build you can run colcon by running the build-ros2.sh script:
+
+.. code-block:: bash
+
+    ./build-ros2.sh
```
Comment (carlossvg) [inline source/Installation/Rolling/QNX-Development-Setup.rst:316]: I would add a section briefly explaining which rmw implementations are compatible with this distro and this version. Or at least which ones have been tested.

File: source/Installation/Rolling/QNX-Development-Setup.rst (hunk 2 above)
Comment (asobhy-qnx) [reply source/Installation/Rolling/QNX-Development-Setup.rst:121]: https://raw.githubusercontent.com/ros2/ros2/rolling/ros2.repos does not exist. This is for rolling https://raw.githubusercontent.com/ros2/ros2/master/ros2.repos

**Comment by asobhy-qnx** [comment]
@landerU you are correct about cyclonedds version change. This commit broke the build https://github.com/ros2/rmw_cyclonedds/commit/9035d1f3d1535e7bd47b3340bfe8dbfc68cad379#diff-2c117cdb07f2617046f7f9886eb3fdd3466ca9a5df25213a5edc4eb34af7f1d6

The issue is that I am patching cyclonedds to build for QNX and I have to patch a specific tag otherwise the patch will not work as well.

I will update it manually for now until I can figure a solution.

**Comment by LanderU** [comment]
@asobhy-qnx, this is why I'm suggesting to use a CI, to ensure that everything still works, and if not, to fix it.

File: source/Installation/Rolling/QNX-Development-Setup.rst [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. include:: ../_Apt-Repositories.rst
+
+Install development tools and ROS tools
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. code-block:: bash
+
+   sudo apt update && sudo apt install -y \
```
[the remark this answers is not in the document]
Comment (carlossvg) [reply source/Installation/Rolling/QNX-Development-Setup.rst:63]: Yes, I think that's easier

File: source/Installation/Rolling/QNX-Development-Setup.rst (hunk 2 above)
Comment (carlossvg) [reply source/Installation/Rolling/QNX-Development-Setup.rst:121]: Ah, you're right :)

**Comment by asobhy-qnx** [comment]
@LanderU How can I setup CI on gitlab?

File: source/Installation/Rolling/QNX-Development-Setup.rst [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+      rsync \
+      rename
+
+    python3 -m pip install -U \
+      Cython \
+      numpy \
+      lark-parser
+
+    # Optional: If CycloneDDS is needed then it has to be build for host first to use dssconf tool required when cross compiling
+    sudo apt install bison
```
Comment (LanderU) [inline source/Installation/Rolling/QNX-Development-Setup.rst:110]:
```suggestion
    sudo apt install -y bison
```

**Comment by LanderU** [comment]
@asobhy-qnx, I'll fork your repo and add a basic CI draft.

**Comment by carlossvg** [comment]
> @asobhy-qnx We should add the document reference to the install instructions index

I would add QNX instructions reference here: https://github.com/ros2/ros2_documentation/blob/rolling/source/Installation.rst#installation

I see that this document is different in rolling than the one in master. Reading the comment in your fork it looks like you should forking rolling instead of master (https://github.com/asobhy-qnx/ros2_documentation/blob/master/source/Installation.rst). 

@clalancette Could you clarify this and suggest how to add the QNX document to the Installation index?

**Comment by asobhy-qnx** [comment]
> @carlossvg looks like you should forking rolling instead of master

Thanks for catching that. After we are done reviewing here, I will create another PR to merge to rolling.

**Comment by clalancette** [comment]
> @clalancette Could you clarify this and suggest how to add the QNX document to the Installation index?

Apologies for the delay here.

My suggestion for the time being is that we actually host the QNX documentation elsewhere (github or wherever you think is best), and then put a link to it from https://docs.ros.org/en/rolling/Related-Projects.html .  If and when QNX becomes a Tier 2 or Tier 3 platform, we can then think about moving it into the standard Installation place.

**Comment by asobhy-qnx** [comment]
it could be placed here https://github.com/asobhy-qnx/ros2_documentation/tree/rolling/source/Related-Projects

**Comment by carlossvg** [comment]
@asobhy-qnx I understand that @clalancette is suggesting to add just the link in the original ros2 documentation repo, here https://github.com/ros2/ros2_documentation/blob/rolling/source/Related-Projects/Intel-ROS2-Projects.rst and host the instructions in an external place. I see two good options to host the instructions in the meanwhile:

1. Link instruction document in this fork
   - https://github.com/asobhy-qnx/ros2_documentation/blob/master/source/Installation/Rolling/QNX-Development-Setup.rst
2. Put the instructions in the ROS 2 real-time working group documentation 
   - https://real-time-working-group.readthedocs.io/en/latest/Guides/Real-Time-Operating-System-Setup/QNX/build_qnx.html

Option 1 is probably better because you own the repo and you can update the instructions easily. In this case, I would suggest you use [readthedocs](https://readthedocs.org/) to host the docs of this fork properly formated instead of just linking the `.rst` file directly.

We could support you in creating the documentation hosting, it's pretty straightforward as far as I know.  

cc @LanderU

**Comment by asobhy-qnx** [comment]
Thank you @carlossvg I do appreciate your support. I would have to get approval first before I do anything since this current location was supposed to be temporary until we merge into ROS2.

**Comment by LanderU** [comment]
Yes, if you decide to use readthedocs and you need support for doing so, I can help you in the process.

**Comment by asobhy-qnx** [comment]
Thank you @LanderU will let you know

**Comment by asobhy-qnx** [comment]
Hello @carlossvg , @LanderU and @clalancette,

I created a new page for ros2 qnx docs here https://ros2-qnx-documentation.readthedocs.io. Also I moved the docs repo to https://gitlab.com/qnx/ros2/docs.

I created CI builds for foxy, galactic and rolling here: https://gitlab.com/qnx/ros2/ros2_qnx/-/pipelines

Thank you all for your support and advice. As always if you have any recommendations or questions please let me know.

Also please feel free to provide links for https://ros2-qnx-documentation.readthedocs.io in your own docs.

Thank you,
Ahmed

**Comment by LanderU** [comment]
Thanks for the ping @asobhy-qnx, this is awesome!

**Pull request merged**